[Enhancement] Matlab-like type system

balbasty opened the pull request:
Implementation of #20 

I am leaving this as a draft PR for now.

Todo:
- [x] Unit tests for matlab-like indexing.
- [x] Unit tests for (nested) conversion to/from runtime objects.

~The `MatlabClassWrapper` test fails on my side, but I don't think I've modified that class.~ (fixed)

johmedr reviewed (commented):
Thank you for implementing all these changes @balbasty -- I think this is definitely a big step forward for our type system! 

I left a few comments in the review, in particular about attaching helper functions as static class methods to the class they belong to, to keep a clear namespace for people doing `from spm import *`. I think the main implementations (e.g., for converting a cell to a num array) should be attached to  the class it belong (e.g., `Array.from_cell(...)` or `Cell.as_array()`), with the possibility of adding helper wrappers for users (e.g., `num2cell`) that just rely on this implementation and that can be imported selectively (e.g., `from spm.helpers import *`). This would keep ease of use for Matlab users but allow for more Python-like syntaxes, e.g. 
```python
my_cell.as_array().as_type(dtype=np.uint8).reshape((4,3))
```

The second point is that we need to test that this type systems actually interfaces from and to Matlab without issues. The tests do not test for that yet. A simple check is to construct an identity function and test for equality: 
```python 
idt = Runtime.call('eval',  '@(x) x')
assert idt(s) == s
``` 
I think we need to get this into the tests before approving the PR. 

Other than this, everything looks good, thank you for this :)

johmedr commented:
Following up on the `asmatlab` function, doing things like `Runtime.call('eval', '@(x) x')` does not work as it returns a function handle which is not handled by `asmatlab`. I would suggest that we handle unknown objects by simply returning them to the user. After all, if Matlab Runtime did not complain about converting through, never should we!

balbasty commented on spm/__wrapper__.py:
```diff
@@ -14,27 +14,1095 @@
+# ----------------------------------------------------------------------
+
+
+def cell(*iterable):
```
Current constructors do indeed follow the `__init__(self, shape_or_iterable, ...)` syntax. However, it leaves room to ill-definite cases:
- `Cell([1, 2, 3])` generates a cell-array of size `1x2x3` so, how to create a cell that contains `1, 2, 3` -- i.e. `{1, 2, 3}` in matlab?
- `Cell([[3, 4], [5, 6]])` generates a cell-array of size `2x2` that contains the values `3, 4, 5, 6`, so how to create a cell of cells -- i.e. `{{3, 4}, {5, 6}}` in matlab?

So the `cell(...)` helper (which resolves these issues) should probably become the  `Cell.from_iterable(iterable)` class method that you propose.

Alternatively, we could have specific methods to construct cell/struct/num arrays of a given shape, like `Cell.from_shape(shape)` or `Cell.empty(shape)`, and always trigger a copy-construction when using the `Cell(array_like)` syntax. But it deviates a bit from the matlab syntax, so I am not a fan.

Do we agree with that we should implement the `Cell.from_iterable(iterable)` syntax?

johmedr replied:
I think this is a really good point ! 
> Do we agree with that we should implement the `Cell.from_iterable(iterable)` syntax?
Yes, that's my preferred option, if there is no strong arguments against it.

balbasty commented on spm/__wrapper__.py:
```diff
@@ -14,27 +14,1095 @@
+    return obj
+
+
+def num2cell(array):
```
Sounds good!

balbasty commented on spm/__wrapper__.py:
```diff
@@ -14,27 +14,1095 @@
+    return asmatlab(Struct(other))
+
+
+def asmatlab(other):
```
My thinking was that some users may want to convert python/json structures to our types, e.g.

- `asmatlab([{"a": 1}, [0, 1, 2]])` returns a `Cell` that contains a `Struct` and an `Array`.

But I agree that may not be needed (especially as users can directly pass such a structure to an `spm_*` function without needing the explicit conversion).

balbasty commented on spm/__wrapper__.py:
```diff
@@ -14,27 +14,1095 @@
+            obj = obj.reshape(shape)
+            return obj.tolist()
+        return obj
+        # TODO: what about sparse numpy arrays? Does matlab understand them?
```
OK so if the user passes a sparse numpy array as input, we should convert it to a dict with `type__="sparse"`?
Do you have a documentation about the expected layout/content of the `data__` field in this case?

johmedr replied:
Currently, sparse arrays are converted to dense arrays and passed in the `data__` field, we could find a more efficient way to handle this in the future.

balbasty commented on spm/__wrapper__.py:
```diff
@@ -14,27 +14,1095 @@
+    if isinstance(other, Array):
+        return other
+
+    if isinstance(other, Cell):
```
I think there's a misunderstanding about the use of the `asmatlab` function. It does not convert objects to representations that we can pass to the runtime -- I use the `_as_matlab_object` for this purpose.

Instead, `asmatlab` converts python objects to our type system (recursively). 
Hence why this function should be renamed.

Essentially there are two functions
- One that converts python objects to "Our types"
- One that converts python objects to "Matlab-compatible types"

| Matlab | Matlab-compatible types | Our types |
| - | - | - |
| `Nx...  {single, double, ...}` | `{ndarray, matlab.single, matlab.double, ...}` | `Array(N, ...)` |
| `1x1 {single, double, ...}` | `{double, int, complex, bool}` | `Array()` |
| `1xN cell` | `{list, tuple, set}` | `Cell(N)` |
| `NxMx... cell` | `dict(type__="cellarray")` | `Cell(N, M, ...)` |
| `1x1 struct` | `dict` | `Struct()` |
| `Nx... struct` | `dict(type__="structarray")` | `Struct(N, ...)` |

Is that it?

balbasty commented on spm/__wrapper__.py:
```diff
@@ -14,27 +14,1095 @@
+    #       * we should hide all math functions.
+    #       * `__add__`, `__iadd__`, `__radd__` should fallback to `extend()`,
+    #         as in tuples, rather than `np.add` as in arrays.
+    #       * "scalar" cells should be forbidden (already the case at
```
I think we agree. A `Cell` with a zero or one elements exist, but they must have at least one dimension -- their shape must be `(0,)` or `(1,)`. What I call a "scalar cell" (and think should be forbidden) is a `Cell` whose shape is an empty `tuple()`, which could be created by doing something like `Cell(1).reshape([])`.

In contrast, "scalar" `Array` and `Struct`  are allowed:
- a "scalar array" is the same as e.g. `np.asarray(1.0)`, which is identical to the python scalar `1.0` for all purposes. 
- a "scalar struct" is the same as a dictionary (with dot access to fields)

johmedr replied:
I see, I had the wrong idea of what a scalar cell is. This sounds good!

balbasty commented:
Thanks for the thotough review @johmedr! I've answered to a few comments. I should have some time next week to implement your suggested changes.

It would be nice to have the identity function in the bindings, indeed.
Also, maybe implement a few functions that either return known objects, or expect known objects and check their value?

Or, can we call pure matlab function through the binding? Like can I call `Runtime.instance().mpython_endpoint("cell", (2, 1))`?

johmedr commented:
> Thanks for the thorough review @johmedr! I've answered to a few comments. I should have some time next week to implement your suggested changes.

Great, thank you! I'll try to update the tests asap to go pass data to the Matlab Runtime and back. 

> It would be nice to have the identity function in the bindings, indeed. Also, maybe implement a few functions that either return known objects, or expect known objects and check their value?
> 
> Or, can we call pure matlab function through the binding? Like can I call `Runtime.instance().mpython_endpoint("cell", (2, 1))`?

We can call `Runtime.call`, which is basically `feval` with conversion of argument types. Calling `Runtime.instance().mpython_endpoint` bypasses the type conversion on the Python side but not on the Matlab side (that's what `mpython_endpoint` is essentially doing).

balbasty commented:
The type system is now in close-to-final form, but needs many more unit tests.

One final important point to deal: currently, `Cell(list[list])` returns a Cell array, but lists are used by matlab to indicate 1D cells, so a `list[list]` should really be a `Cell[Cell]`. It means I need to _not_ use `asanyarray` in the cell constructor. And we also probably need a method to transform a `Cell[Cell]` into a deep cell array. Maybe `deepcat()`?

I have renamed quite a few methods. @johmedr can you please go through the code and check that the API suits you?

The final thing that is not renamed is `MatlabClassWrapper`, which I would vote to rename `MatlabClass`.

balbasty commented:
One of the tests fails (but was not failing locally). It seems the `np.ndarray.data` field of a struct array is accessed, but it is hidden, so the Struct thinks it is a new key that must be defined and adds it to its dictionary. Not sure where that call to `.data` happens.

balbasty commented:
Also, I've added `check_finalized` in delayed arrays, so that if a delayed array that is already finalized is accessed, an error is raised. This should cover the "bad" use cases you mentioned during the meeting.

balbasty commented:
It's starting to look good. I've put a couple of questions/comments in the code, but it would be good of someone else than me tried it now :)

johmedr commented:
Hi @balbasty, I was playing with getting "DelayedArrays" without setting them, and found an error that allows to transform a DelayedArray to an empty object that then would be propagated to Matlab: 
```python 
>>> s = Struct()
>>> s[1].foo(4)
<spm.__wrapper__.DelayedArray object at 0x175e16210>
>>> s
Struct([{},
        {'new': Cell([<spm.__wrapper__.DelayedArray object at 0x1759769c0>,
              <spm.__wrapper__.DelayedArray object at 0x175e15a00>,
              <spm.__wrapper__.DelayedArray object at 0x175e17410>,
              <spm.__wrapper__.DelayedArray object at 0x175e16390>,
              <spm.__wrapper__.DelayedArray object at 0x175e16210>])}      ])
>>> s.values()
dict_values([Cell([Cell([]),
      Cell([Array([]), Array([]), Array([]), Array([]), Array([])])])])
>>> s
Struct([{},
        {'new': Cell([Array([]), Array([]), Array([]), Array([]), Array([])])}])
```
Any idea of what could be happening there?

balbasty commented:
The way it is currently setup, as soon as one DelayedArray is finalized (i.e., knows its final type), it tells its parent which then forces all of its element to be finalized. If a `DelayedArray` is not yet finalized by that time, it is transformed into a "default value", which is the empty array if the parent is a `Cell`.

I did not foresee issues with doing it that way (at least not when writing SPM batches the typical way). But we could switch it so that these unfinalized arrays are kept as `DelayedArray` for longer?

The use case I had is that if I do this in matlab:
```matlab
s.c(3).f = 'a';
```
then `c{0}` and `c{1}` contain empty arrays.

If I do something similar in python:
```python
s = Struct()
s.c(2).f = 3 
```
- By the time I do `s.c`, `c` is a DelayedArray.
- By the time I do `s.c.__call__`, `c` knows that its final type is `Cell`
- By the time I do `s.c(2)`, `c` is a `Cell`of size `[3]` containing 3 `DelayedArray`s
- By the time I do `s.c(2).d`, `c[2]` knows it is a Struct, finalizes itself and tells so to its parent.
- `c` can then replace all of its non final elements with empty arrays
- and we have the same final result as in Matlab.

Another solution could be that when `c(2)` is called and `c` is extended so that it becomes of length `3`, only the indexed element (here, `2`) is assigned a `DelayedArray`, and non indexed elements are already set to their default value (= empty arrays).

johmedr commented:
> Note that that doesn't explain why a call to `s.values()` generates empty arrays. And also why is there a `new` key in one of the structs...

My bad, I used `new` instead of `foo` in my tests but did not update the results --- edited now.

johmedr commented:
I understand your point, thank you for the detailed example. 

> I did not foresee issues with doing it that way (at least not when writing SPM batches the typical way).

I agree with you, my concern is more about getting obscure errors while writing the batch, e.g. when initializing fields in a loop. In particular, the batch system can be quite picky and tricky to get right on the first time ...     

Do you think there would be a way to Delay the finalization of the Array? For instance, a delayed array could have a future type (`_future_self = ...`), which is immutable. However, the transformation from DelayedArray to an existing type is triggered only if the terminal object (or any object down the line) is finalised explicitly. I understand that this could cause problem wrt the type of indexing operators that are allowed, but maybe DelayedArray could try the requested indexing operation on its future type to decide how to behave.  

In addition, it could be helpful that parents don't hold references to their uninitialised children. That way, if anything breaks down the indexing chain, the uninitialised part becomes an orphan and is removed by the garbage collector (unless explicitly affected to a variable, but that's not our problem).

That would give the following behaviors:

```python
>>> s = Struct()
>>> s.c(2).f = 3      # OK
```
- By the time I do `s.c`, `c` is a DelayedArray with `s.c._promised_type = None`. `s` does not have yet a reference to `c`
- By the time I do `s.c.__call__`, `c` knows that its final type is Cell (e.g., `s.c._future_self = Cell()`)
- By the time I do `s.c(2)`, `c` is a DelayedArray with `s.c._future_self = Cell(3)`
- By the time I do `s.c(2).__setattr__('f', 3)`, `c[2]` knows it is a `Struct` (`s.c._future_self = Struct`), and one of its field was finalized, so it finalize the Struct and its parents 
- Finally, a reference to `c` is added to `s`


Now, when we do: 

```python 
>>> s = Struct()
>>> s.c(2).f 
AttributeError: 'DelayedArray' has no attribute 'f'
>>> s.c
AttributeError: 'Struct' has no attribute 'c'
```
- By the time I do `s.c`, `c` is a DelayedArray with `s.c._future_self = None`. `s` does not have yet a reference to `c`
- By the time I do `s.c.__call__`, `c` knows that its final type is Cell (e.g., `s.c._future_self = Cell`)
- By the time I do `s.c(2)`, `c` is a DelayedArray with `s.c._future_self = Cell(3)`
- By the time I do `s.c(2).f`, `c[2]` knows it is a `Struct` (`s.c._future_self = Struct(3)`), returns a DelayedArray
- By the time I do `s.c(2).f.__repr__()`, I try to call an unallowed function on a DelayedArray.
- it makes the chain collapse: all parents of `f` that are DelayedArrays are left uninitialised or removed (e.g., delete the references to them and possibly call garbage collector)  
- `s` does not have a reference to `c`, that chain is an orphan and will be removed by the garbage collector

What do you think?

johmedr commented:
On another note, because cells are based on numpy array, they reproduce their `__eq__` behaviour. This leads to confusing logic: 
```python 
>>> np.zeros(0) == np.zeros(0)
np.array([])
>>> bool(np.zeros(0) == np.zeros(0)) 
False
>>> bool((np.zeros(0) == np.zeros(0)).all())
True
``` 

This naturally propagates to containers that are agnostic to that `.all` logic:

```python
>>> a = {'a': np.zeros(0)}
>>> b = {'a': np.zeros(0)}
>>> a == b
False
```

In our case, we are facing the same behaviour: 
```python 
>>> a = Cell()
>>> b = Cell()
>>> a == b
False
```

And thus: 
```python 
>>> s1 = Struct() 
>>> s1.a = Cell()
>>> s2 = Struct()
>>> s2.a = Cell()
>>> s1.a == s2.b 
False
```

Similarly, 
```python 
>>> s1 = Struct() 
>>> s1.a = Cell.from_array([1, 2])
>>> s2 = Struct()
>>> s2.a = Cell.from_array([1, 2])
>>> s1 == s2
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

In Matlab, the '==' operator is undefined for both struct and cell arrays. We are free to define the behaviour that suits us best. Do you think we should attempt `x.all` when `bool(x)` raises a ValueError?

balbasty commented:
Hooo I like that!!! (the promise logic). It's true that the creation of a Delayed Array should always end with a __setitem__ or __setatt__. If it does not, it's forbidden syntax. I think that + initializing non-indexed implictely-created elements with `[]` rather than a DelayedArray will cover all bad cases.

How do you insert the finalized back in the parent, if the parent does not hold a reference? The Delayed array holds a reference to the parent _and_ the index in which it must insert itself?

Re the conflicting behaviour between array/list/dict, I completely agree. I went back and forth on this several times. I agree that in the end Struct and Cell should probably mimic the list and dict behaviour exactly (and maybe also inherit from list/dict, so that downstream applciation that test `isinstance` can use Cells and Structs directly). I am not sure we lose much be not having "element wise" operations on cells and structs.

balbasty commented:
We should also forbid the assignment of a DelayedArray into anything (unless we are the ones doing it behind the scenes).

johmedr commented:
> Hooo I like that!!! (the promise logic). It's true that the creation of a Delayed Array should always end with a `__setitem__` or `__setattr__`. If it does not, it's forbidden syntax. I think that + initializing non-indexed implictely-created elements with `[]` rather than a DelayedArray will cover all bad cases.

Yes exactly. 

> How do you insert the finalized back in the parent, if the parent does not hold a reference? The Delayed array holds a reference to the parent and the index in which it must insert itself?

That's a good question. This error case aims to be an edge case  (i.e., it aims at raising an error to inform the user to update their syntax), thus we can afford to spend a bit of time asking the parent to find and dereference its child (e.g., parent._dereference(self) or something like this). Otherwise, we can just have two internal storage in the parents, one for the proposed indexing, holding reference to the unfinalised child, and one for the finalised child. Upon finalisation, we always clear the secondary structure. Could be something simple like `self._promised_child = {index_or_key: child_ref}`.  

> We should also forbid the assignment of a DelayedArray into anything (unless we are the ones doing it behind the scenes).

Agree with this, haven't found an easy way to do it yet. The only solution I see would be to make it an abstract class, raising  an error upon initialisation. Do you have any solution in mind?

balbasty commented:
I've now implemented `__bool__` (lowered to `all()`) on our arrays.

One weird behaviour is that this comparison broadcasts, so `bool(Cell.from_any([1, 1, 1]) == 1) is True`, which differs from  `list`.

One other difference is that a `list` can only be compared to another `list`: `([1, 2, 3] == (1, 2, 3)) is False`.
Whereas `bool(Cell.from_any([1, 2, 3]) == [1, 2, 3]) is True` and  `bool(Cell.from_any([1, 2, 3]) == (1, 2, 3)) is True`.

johmedr commented:
@balbasty, I was wondering whether we should make sure that unidimensional cells and arrays are 2-dim, column-based (e.g., Cell(1,3)). This is to ensure that types and shapes are consistent between Matlab and Python. For now, 
``` python
>>> Runtime.call('cell', 3, 1)
Cell([Cell([]), Cell([]), Cell([])]) # This should be a Cell of empty Arrays? 
>>> Runtime.eval('cell', 3, 1).shape # This should be (3, 1)
(3,) 

>>> Runtime.call('cell', 1, 3)
Cell([Cell([]), Cell([]), Cell([])]) # This should be a Cell of empty Arrays? 
>>> Runtime.call('cell', 1, 3).shape # This should be (1, 3)
(3,)

>>> Cell(3, 1)
Cell([[Array([])],
      [Array([])],
      [Array([])]])
>>> Cell(3, 1).shape
(3,1)

>>> Cell(1, 3)
Cell([[Array([]), Array([]), Array([])]])
>>> Cell(1, 3).shape
(1, 3)

>>> Cell(3)
Cell([Array([]), Array([]), Array([])])
>>> Cell(3).shape  # This should be (1, 3)
(3,)
```
I am adding tests to ensure this. 

What do you think?

johmedr commented:
I understand the indexing issue. I think the question is more whether some functions may fail because they expect a 3x1 cell array instead of a 1x3 cell array. Do you could think we could just internally keep track of the Matlab shape of the array if ndim = 1? or more easily, have the array be `(1,n)` or `(n,1)` and call 
```python 
if np.prod(self) == np.max(np.shape(self)): 
    indices = np.unravel_index(indices, np.shape(self), order='F')    

super().__getitem__(indices)
```

balbasty commented:
Currently, shapes `(N,)` and `(1, N)` both get converted to a `1xN` cell. And only shape `(N, 1)` gets converted to a `Nx1` cell. Is that not good enough? And in the other direction I think a `1xN` cell gets converted to a `(N,)`-shaped Cell (??).

I am a bit worried of changing the indexing behaviour from standard numpy, just because that might throw users (who are used to python) off.

balbasty commented:
To support my point:

```python
>>> Runtime.instance().mpython_endpoint('eval', "cell(1, 3)")
[matlab.double([]), matlab.double([]), matlab.double([])]
```

balbasty commented:
And the issue with this unexpected output
```python
>>> Runtime.call('cell', 3, 1)
Cell([Cell([]), Cell([]), Cell([])]) # This should be a Cell of empty Arrays? 
```

Is that in the Cell constructor I recursively convert the cell's content in a not very smart way. I'll fix that!

johmedr commented:
> Currently, shapes `(N,)` and `(1, N)` both get converted to a `1xN` cell. And only shape `(N, 1)` gets converted to a `Nx1` cell. Is that not good enough? And in the other direction I think a `1xN` cell gets converted to a `(N,)`-shaped Cell (??).

I am not sure to follow here, do you mean when passed to Matlab? 

> To support my point:
> 
> ```python
> >>> Runtime.instance().mpython_endpoint('eval', "cell(1, 3)")
> [matlab.double([]), matlab.double([]), matlab.double([])]
> ```

I agree with this, though  we could work around this by editing `mpython_endpoint` to systematically report the shape of Cells. Places where it can be useful is in DCM: PEB uses Cell arrays to store models per subject, so a (1xN) cell array has one subject and N models (and doesn't make sense because group studies on 1 subject don't) and (Nx1) has 1 model and N subject. Here getting the shape right is really important. 
 
> I am a bit worried of changing the indexing behaviour from standard numpy, just because that might throw users (who are used to python) off.

Agreed not to do it, I did not think about that!

balbasty commented:
> I am not sure to follow here, do you mean when passed to Matlab?

Yes, A python `Cell([3])` becomes a matlab `1x3 cell`, and so does a python `Cell([1, 3])`.

> a (1xN) cell array has one subject and N models (and doesn't make sense because group studies on 1 subject don't) and (Nx1) has 1 model and N subjects. Here, getting the shape right is really important.

Don't you think that properly documenting that a list (or 1D cell) correspond to a `1xN` matlab cell will be enough? The issue you mention only arises if someone specifies their model by doing
```
x.y = [a, b, c]
```
thinking that they are defining a column vector when in reality they are defining a row vector. They should do 
```
x.y = Cell([a, b, c])[:, None]
```
which I agree is ugly, or
```
x.y = Cell([[a, b, c]]).T
```
which is the syntax I used in [DCM_ERP](https://github.com/spm/spm-python/blob/b5966a83d66b77bf0bfbe867bd2ca12b9afcfa45/examples/DCM_ERP.py#L95).

johmedr commented:
> Yes, A python `Cell([3])` becomes a matlab `1x3 cell`, and so does a python `Cell([1, 3])`.
> 
> 

We can document that for now and move on, I agree that it should not be a big deal. 

I think I am concerned with things that might occur when interchanging between Matlab and Python, e.g.: 
- I saved my cell array (Nx1) to a mat file
- I want to load it in Python and edit some of its elements
- I want to pass it to `spm_dcm_peb` or similar, which expects a Nx1 cell array. 

I have the feeling that step 1 would yield an `(N, )`-shaped cell, which would then need to be explicitly converted (e.g., `gcm.reshape((-1, 1))`) before calling `spm_dcm_peb`.  It is not terrifying per se... But it requires me to know what shape `spm_dcm_peb` expects as an argument. This is fine in this case but when it happens to e.g. a hidden and undocumented field in the deeps of SPM, we might have a hard time diagnosing the error! 

My rationale is that the shape of cell arrays might contain information, and the fact that one or the other dimension is 1 might be really important. Because we cannot rule this case out, we should make sure this information is not altered between Python and Matlab. I think that Cell(N) creating a (1xN) cell when passed to Matlab is fined, but `Runtime.call('eval', '{1;2;3}')` and `Runtime.call('eval', '{1,2,3}')` should not be equal IMO.

johmedr commented:
Okay, I am recompiling spm-python, the new `mpython_endpoint` will have the following behavior wrt cells: 

```matlab
>> c = mpython_endpoint('cell', 1, 2)

c =

  1×2 cell array

    {0×0 double}    {0×0 double}

>> c = mpython_endpoint('cell', 2, 1)

c = 

  struct with fields:

    type__: 'cell'
    size__: [2 1]
    data__: {[]  []}
``` 

Should save us trouble, now we enforce the equivalence `Cell(N) == Cell(1,N) == Runtime.call('cell', 1, N) != Cell(N,1)`

balbasty commented:
Hi man ok I finally understand, sorry for eing slow!

I was convinced that column vectors where passed as cell arrays (`{"type__": "cell", ...}`). I completely agree that column vectors should be passed as Nx1 cell arrays, not as simple lists, to differentiate them from row arrays. Your fixes look good to me.

johmedr commented:
Also, it seems that flat iterators on struct are infinite, I don't really now what is happening. Ran into this issue already?

balbasty commented:
I had the same issue with __iter__ which I had to overload. I think it's because numpy relies on IndexError to stop its iteration. But we don't have them anymore thanks to implicit resizing 😅

johmedr commented:
Here is what I get: 


```python 
>>> Struct(1).flat # stuck
>>> [*Struct(1).flat] # stuck
>>> s = Struct()
>>> s.foo = 'bar' 
>>> s.flat
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/johan/Documents/Python/spm-python/spm/__wrapper__.py", line 551, in _error_is_not_finalized
    raise IndexOrKeyOrAttributeError(
spm.__wrapper__.IndexOrKeyOrAttributeError: 'This DelayedArray has not been finalized, and you are attempting to use it in a way that may break its finalization cycle.\nIt most likely means that you are indexing out-of-bounds without *setting* the out-of-bound value.\n* Correct usage:   a.b(i).c = x\n* Incorrect usage: x = a.b(i).c\n'
>>> [*s.flat] # stuck
```

balbasty commented:
Ha that's different. I did not include "flat" as one of the authorised methods or properties in a Struct. I tried to keep them at a minimum so that they can be used as keys instead.

If you want to add flat to the list, you can add it to the class variable defined at the beginning of Struct.

johmedr commented:
Updating tests. 
Cells have one thing left to fix: 

```python
>>> Runtime.call('cell', 1,3)
Cell([Cell([]), Cell([]), Cell([])]) 
```
should be 
```python
>>> Runtime.call('cell', 1,3)
Cell([Array([]), Array([]), Array([])]) 
```
and same for `cell(3,1)`

balbasty commented:
@johmedr I believe I have fixed everything.

I had to change some tests (mostly replaced 1D indexing of column vectors with 2D indexing). Please check that you're fine with their behaviour.

One discrepency remaining is that `Runtime.call("cell", 1, 3)` returns a cell of empty arrays, where each empty array has shape `[0, 0]`, whereas I use empty arrays of shape `[0]` when defining default elements on the python side. Should I use `[0, 0]` to better mimic matlab?

johmedr commented:
Amazing, thank you so much ! Happy with the change you made. 

> One discrepency remaining is that Runtime.call("cell", 1, 3) returns a cell of empty arrays, where each empty array has shape [0, 0], whereas I use empty arrays of shape [0] when defining default elements on the python side. Should I use [0, 0] to better mimic matlab?

Yes, I think that makes sense (for consistency). 
Are we ready to merge after this? 😄

balbasty commented:
- Fixed empty array shape `(0,) -> (0, 0)`
- Used new way of (de)serializing sparse arrays
- Changed sparse layout from COO to CRC + fixed (de)serialization

One (last?) inconsistency:
```python
>>> Runtime.call('cell', 1, 3).shape
(3,)

>>> Runtime.call('zeros', 1, 3).shape
(1, 3)

>>> Runtime.call('struct', 'a', [1, 2, 3], 'b', [4, 5, 6]).shape
(1, 3)
```

So in essence, a row cell is converted to 1D `Cell` with shape `(N,)`, whereas a row array is converted to a 2D `Array` with shape `(1, N)` and a row struct array is converted to a 2D `Struct` with shape `(1, N)`.

Should we always convert row arrays to 1D things python side?

johmedr commented:
> * Fixed empty array shape `(0,) -> (0, 0)`
> * Used new way of (de)serializing sparse arrays
> * Changed sparse layout from COO to CRC + fixed (de)serialization
Amazing! 

> One (last?) inconsistency:
> 
> ```python
> >>> Runtime.call('cell', 1, 3).shape
> (3,)
> 
> >>> Runtime.call('zeros', 1, 3).shape
> (1, 3)
> 
> >>> Runtime.call('struct', 'a', [1, 2, 3], 'b', [4, 5, 6]).shape
> (1, 3)
> ```
> 
> So in essence, a row cell is converted to 1D `Cell` with shape `(N,)`, whereas a row array is converted to a 2D `Array` with shape `(1, N)` and a row struct array is converted to a 2D `Struct` with shape `(1, N)`.
> 
> Should we always convert row arrays to 1D things python side?

That could be nice to squeeze the first axis of 2d Arrays and Structs so that indexing is consistent, e.g., 
```python
>>>  s = Runtime.call('struct', 'a', [1, 2, 3], 'b', [4, 5, 6])
>>>  s[1].b
5
```
instead of 
```python
>>>  s[0, 1].b
5
```

 
> And let me paste comments from the code so that they are not forgotten:
> 
> * MATLAB does not have 1D arrays (they are always 2D).
>   A 1D python array is interpreted as a row arrays, so the round trip
>   goes [1, 2, 3] -> [[1, 2, 3]].
>   Are we happy with that? I think it works in the sense that
>   when a numpy operation takes a matrix and a vector, the vector is
>   broadcasted to its left, and is therefore interpreted as a row vector.
>   !! We should clearly document this behaviour.
> * The creation of numeric vectors on the python side is currently
>   quite verbose (`Array.from_any([0, 0])`, because `Array([0, 0])`
>   is interpreted as "create an empty array with shape [0, 0]").
>   We could either
>   
>   * introduce a concise helper (e.g., `num`) to make this less verbose:
>     `Array.from_any([0, 0])` -> `num([0, 0])`

Makes sense. 

>   * Interpret lists of numbers as Arrays rather than Cells. But this is
>     problematic when parsing the output of a mpython_endpoint call, since
>     lists of numbers do mean "cell" in this context.

The problem is that we don't have much freedom with this, see Matlab doc [here](https://www.mathworks.com/help/matlab/matlab_external/pass-data-to-matlab-from-python.html) and [here](https://www.mathworks.com/help/matlab/matlab_external/handle-data-returned-from-matlab-to-python.html). 

> * I've added support for "object arrays" (such as nifti or cfg_dep)
>   in DelayedArray:
>   `a.b[0] = nifti("path")` means that `a.b` contains a 1x1 nifti object.
>   However, I only support 1x1 object, and the index must be 0 or -1.
>   There might be a way to make this more generic, but it needs more thinking.
>   The 1x1 case is all we need for batch jobs (it's required when building
>   jobs with dependencies).
>   It might be useful to have a `ObjectArray` type (with `MatlabClass`
>   as a base class?) for such objects -- It'll help with the logic in
>   delayed arrays. It should be detectable by looking for `class(struct(...))`
>   in the constructor when parsing the matlab code, although there are
>   cases where the struct is created beforehand, e.g.:
>   https://github.com/spm/spm/blob/main/%40nifti/nifti.m#L12
>   Maybe there's a programmatic way in matlab to detect if a class is
>   a pure object or an object array? It seems that old-school classes
>   that use the class(struct) constructor are always object arrays.
>   With new-style classes, object arrays can be constructed after the
>   fact:
>   https://uk.mathworks.com/help/matlab/matlab_oop/creating-object-arrays.html
>   After more thinking, it also means that we have again a difference in bhv
>   between `x{1} = object` and `x(1) = object`. In the former case, x is
>   a cell that contains an object, whereas in the latter x is a 1x1 object
>   array.

We could use `isscalar` to detect if an object is an array, and pass its shape in that case. I add this as an issue.

> * We should probably implement a helper to convert matlab batches into
>   python batches.

Agreed. In a first time, if you `Runtime.call('load', 'batch.mat')` you should have a Python version of your Matlab batch.

balbasty commented:
OK so the only thing that's blocking the merge is squeezing the first axis of 2d arrays and structs.

Do you want me to do it python side (in `_from_runtime()`) or should it be done matlab-side, in `mpython_endpoint()`.
The former is probably easier.

balbasty commented:
I've done it python side. It's actually quite nice that `call("size", a)` now returns `[3, 4, 5]` instead of `[[3, 4, 5]]`.

johmedr commented:
Fantastic! I also think it's better and easier to squeeze on the Python side.

johmedr commented:
All looks good, merging! 

Thanks for this huge development @balbasty!
